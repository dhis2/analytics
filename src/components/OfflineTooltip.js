import { useOnlineStatus } from '@dhis2/app-runtime'
import i18n from '@dhis2/d2-i18n'
import { Tooltip } from '@dhis2/ui'
import cx from 'classnames'
import PropTypes from 'prop-types'
import React from 'react'
import { styles } from './styles/OfflineTooltip.style.js'

const OfflineTooltip = ({
    disabledWhenOffline,
    disabled,
    content,
    children,
}) => {
    const { offline } = useOnlineStatus()

    const notAllowed = disabled || (disabledWhenOffline && offline)

    return (
        <>
            <Tooltip
                content={content || i18n.t('Not available offline')}
                openDelay={200}
                closeDelay={100}
            >
                {({ onMouseOver, onMouseOut, ref }) => (
                    <span
                        className={cx('tooltip', { notAllowed })}
                        onMouseOver={() => notAllowed && onMouseOver()}
                        onMouseOut={() => notAllowed && onMouseOut()}
                        ref={ref}
                    >
                        {children}
                    </span>
                )}
            </Tooltip>
            <style jsx>{styles}</style>
        </>
    )
}

OfflineTooltip.propTypes = {
    children: PropTypes.node,
    content: PropTypes.string,
    disabled: PropTypes.bool,
    disabledWhenOffline: PropTypes.bool,
}

OfflineTooltip.defaultProps = {
    disabled: false,
    disabledWhenOffline: true,
}

export { OfflineTooltip }
