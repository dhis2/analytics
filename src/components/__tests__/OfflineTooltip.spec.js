import { useDhis2ConnectionStatus } from '@dhis2/app-runtime'
// import i18n from '@dhis2/d2-i18n'
import { render, screen, fireEvent } from '@testing-library/react'
import React from 'react'
import '@testing-library/jest-dom/extend-expect'
import { OfflineTooltip } from '../OfflineTooltip.js'

jest.mock('@dhis2/app-runtime', () => ({
    useDhis2ConnectionStatus: jest.fn(),
}))

const mockUseDhis2ConnectionStatus = (isDisconnected) => {
    useDhis2ConnectionStatus.mockReturnValue({ isDisconnected })
}

describe('OfflineTooltip', () => {
    beforeEach(() => {
        jest.clearAllMocks()
    })

    test('renders children without tooltip when online and not disabled', () => {
        mockUseDhis2ConnectionStatus(false)

        render(
            <OfflineTooltip disabledWhenOffline={true} disabled={false}>
                <span>Test Child</span>
            </OfflineTooltip>
        )

        expect(screen.getByText('Test Child')).toBeInTheDocument()
        expect(
            screen.queryByText('Not available offline')
        ).not.toBeInTheDocument()
    })

    test.only('shows tooltip when offline and disabledWhenOffline is true', () => {
        mockUseDhis2ConnectionStatus(true)

        const { container } = render(
            <OfflineTooltip disabledWhenOffline={true} disabled={false}>
                <span>Test Child</span>
            </OfflineTooltip>
        )

        expect(container).toMatchSnapshot()

        // const childElement = screen.getByText('Test Child')
        // fireEvent.mouseOver(childElement)

        // expect(screen.getByText('Not available offline')).toBeInTheDocument()
    })

    test('shows custom tooltip content', () => {
        mockUseDhis2ConnectionStatus(true)

        render(
            <OfflineTooltip
                disabledWhenOffline={true}
                disabled={false}
                content="Custom Tooltip Content"
            >
                <span>Test Child</span>
            </OfflineTooltip>
        )

        const childElement = screen.getByText('Test Child')
        fireEvent.mouseOver(childElement)

        expect(screen.getByText('Custom Tooltip Content')).toBeInTheDocument()
    })

    test('shows tooltip when disabled is true', () => {
        mockUseDhis2ConnectionStatus(false)

        render(
            <OfflineTooltip disabledWhenOffline={false} disabled={true}>
                <span>Test Child</span>
            </OfflineTooltip>
        )

        const childElement = screen.getByText('Test Child')
        fireEvent.mouseOver(childElement)

        expect(screen.getByText('Not available offline')).toBeInTheDocument()
    })

    test('does not show tooltip when disabledWhenOffline and disabled are false', () => {
        mockUseDhis2ConnectionStatus(true)

        render(
            <OfflineTooltip disabledWhenOffline={false} disabled={false}>
                <span>Test Child</span>
            </OfflineTooltip>
        )

        const childElement = screen.getByText('Test Child')
        fireEvent.mouseOver(childElement)

        expect(
            screen.queryByText('Not available offline')
        ).not.toBeInTheDocument()
    })
})
