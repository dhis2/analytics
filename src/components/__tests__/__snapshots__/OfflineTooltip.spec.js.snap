// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OfflineTooltip shows tooltip when offline and disabledWhenOffline is true 1`] = `
<div>
  <span
    class="tooltip notAllowed"
  >
    <span>
      Test Child
    </span>
  </span>
  <style />
</div>
`;
